### ISSUE TITLE
Integration sample with SqlAlchemy ?

### ISSUE THREAD

**Issue by chavz** [pytest-dev/pytest-factoryboy#11]
Hi how would an integration with sqlalchemy works here ??, especially with the `scoped_session`

I currently have this `conftest.py` setup on my flask project:

```
@pytest.fixture(scope='session')
def app(request):
    """Session-wide test `Flask` application."""
    app = create_app('testing')

    # Establish an application context before running the tests.
    ctx = app.app_context()
    ctx.push()

    def teardown():
        ctx.pop()

    request.addfinalizer(teardown)
    return app


@pytest.fixture(scope='session')
def db(app, request):
    """Session-wide test database."""
    def teardown():
        _db.drop_all()

    _db.app = app
    _db.create_all()

    request.addfinalizer(teardown)
    return _db


@pytest.fixture(scope='function')
def session(db, request):
    """Creates a new database session for a test."""
    # connect to the database
    connection = db.engine.connect()
    # begin a non-ORM transaction
    transaction = connection.begin()

    # bind an individual session to the connection
    options = dict(bind=connection, binds={})
    session = db.create_scoped_session(options=options)

    # overload the default session with the session above
    db.session = session

    def teardown():
        session.close()
        # rollback - everything that happened with the
        # session above (including calls to commit())
        # is rolled back.
        transaction.rollback()
        # return connection to the Engine
        connection.close()

    request.addfinalizer(teardown)
    return session

# basically instead of creating and dropping the database for each test
# i just rollback the transaction.
# this setup works perfectly fine without using factory boy
```

I've been searching around the internet on how to make Factory Boy use the same session fixture that pytest use. And then stumble upon this library, but upon looking into the examples, I cant figure how I or this library would do the same.

**Comment by olegpidsadnyi** [comment]
Hi!
If I understand right what you are trying to achieve: You initialize the database in the fixture because you want a special test db?
I don't think you need to do all this db.create_scoped_session(options=options) etc. There's a package that provides SQLAlchemy integration into Flask - https://pythonhosted.org/Flask-SQLAlchemy/ I hope you are using it.
You should have lazy db somewhere globally defined.
db = SQLAlchemy() - Not yet attached to the app.
You can use it to define your models and it won't break the import.

Finally when you create your app you can initialize that db with your app, assuming all db settings are in the application config.

```python

def create_app():
    app = Flask(__name__)
    db.init_app(app)  # <--------HERE
    return app

```

It will create internally all the necessary things like scoped session. See - https://github.com/mitsuhiko/flask-sqlalchemy/blob/master/flask_sqlalchemy/__init__.py#L735

This means you can import your globally defined lazy "db" and use db.session - which is a scoped session (threading local scoped session).

factory-boy has integration with SQLAlchemy. It understands scoped session in the Meta class of the factory. We use our own base class to attach all our factory classes to the right scoped session.

```python

from factory.alchemy import SQLAlchemyModelFactory
from myproject.db import db

class ModelFactory(SQLAlchemyModelFactory):

    """Base model factory."""

    class Meta:
        abstract = True
        sqlalchemy_session = db.session

```

Then you can subclass it and in theory it should work.

There's also a trick with any function that returns session.
Even in your case you still can use current_app accessor from Flask, since your fixture pushes your app into the context.

So you can define some function like

```python

from flask import current_app

def get_my_session():
    return current_app.extensions['sqlalchemy'].db.session

```

Then you can wrap this function into scoped_session and feed it to factory-boy. It will still be lazy, won't break your imports and will let you setup the database connection.

```python

from sqlalchemy.orm.scoping import scoped_session
from factory.alchemy import SQLAlchemyModelFactory

class ModelFactory(SQLAlchemyModelFactory):

    """Base model factory."""

    class Meta:
        abstract = True
        sqlalchemy_session = scoped_session(get_my_session)

```

**Comment by olegpidsadnyi** [comment]
@chavz Did it work for you

**Comment by chavz** [comment]
hello @olegpidsadnyi , sorry took me a while to respond,

The primary reason behind the whole `session` fixture is for me to be able to rollback my transaction for every test and not needing to re-create and drop the whole database every test. For that i followed  [alex's  blog post](http://alexmic.net/flask-sqlalchemy-pytest/), but as mentioned in his blog, it has some caveats, for Flask-SQLAlchemy v2.0 has some issues with it's `SignallingSession`, see [this stackoverflow post](http://stackoverflow.com/a/26624146/2475170) and [pull request 168](https://github.com/mitsuhiko/flask-sqlalchemy/pull/168). To resolve this problem I did alex's suggestion and sub-classed the `SignallingSession`, see code below:

```
# database.py

from flask.ext.sqlalchemy import SQLAlchemy, SignallingSession, SessionBase


class SessionWithBinds(SignallingSession):
    """This extends the flask-sqlalchemy signalling session so that we may
    provide our own 'binds' argument.

    See https://github.com/mitsuhiko/flask-sqlalchemy/pull/168
    Also http://stackoverflow.com/a/26624146/2475170

    """
    def __init__(self, db, autocommit=False, autoflush=True, **options):
        #: The application that this session belongs to.
        self.app = db.get_app()
        self._model_changes = {}
        #: A flag that controls whether this session should keep track of
        #: model modifications.  The default value for this attribute
        #: is set from the ``SQLALCHEMY_TRACK_MODIFICATIONS`` config
        #: key.
        self.emit_modification_signals = \
            self.app.config['SQLALCHEMY_TRACK_MODIFICATIONS']
        bind = options.pop('bind', None) or db.engine

        # Our changes to allow a 'binds' argument
        binds = options.pop('binds', None)
        if binds is None:
            binds = db.get_binds(self.app)

        SessionBase.__init__(
            self, autocommit=autocommit, autoflush=autoflush,
            bind=bind, binds=binds, **options
        )


class TestFriendlySQLAlchemy(SQLAlchemy):
    """For overriding create_session to return our own Session class"""
    def create_session(self, options):
        return SessionWithBinds(self, **options)


db = TestFriendlySQLAlchemy()

```

and then I reference this in my app's `__init__.py`

```
# app/__init__.py
from flask import Flask

from config import config

# instead of using the default `from flask.ext.sqlalchemy import SQLAlchemy()`
# we use an extended version which is more friendlier to tests
from .database import db as _db

db = _db


def create_app(config_name):
    """Create application using the given configuration.

    Application factory which takes as an argument the name of the
    configuration to use for the application. It then returns the
    created application.

    Args:
      config_name (string): name of the configuration.

    Returns:
      Flask: the created application.

    """
    app = Flask(__name__)
    app.config.from_object(config[config_name])
    config[config_name].init_app(app)

    db.init_app(app)  # <--- as you've mentioned. =)
    ...
    return app

```

Although this whole `SignallingSession` problem has been fixed already with [flask-sqlalchemy/pull/249](https://github.com/mitsuhiko/flask-sqlalchemy/pull/249) , it wont probably be released yet untill the whole [2.1 milestone](https://github.com/mitsuhiko/flask-sqlalchemy/milestones/2.1) has been completed.

I tried your first recommendation, but using `pytest.set_trace()`, I can see that the `session` fixture uses a different sqlachemy session instance as with the one factory boy uses (i checked using `model_factory._meta.sqlalchemy_session`).

I also tried your second recommendation, but I always encounter an error with using outside the application/request context. Probably a problem on where I placing the `get_session` function. 

But nevertheless, somehow I was able to make factory boy use the same session instance with my `session` fixture by encapsulating each factory class inside a pytest fixture which injects my `session` fixture. see example below:

```
@pytest.fixture(scope='function')
def model_factory(session):
    Class ModelFactory(factory.alchemy.SQLAlchemyModelFactory):
        Class Meta:
            model = model
            sqlalchemy_session = session
        name = factory.Sequence(lambda n: "Model %d" % n)
   return ModelFactory


def test_session_instance(session, model_factory):
   assert session == model_factory._meta.sqlalchemy_session
```

Don't know though if this is a good solution.

**Comment by thedrow** [comment]
If https://github.com/FactoryBoy/factory_boy/pull/391 is merged we could integrate with pytest-sqlalchemy and provide instances that initialize models from the session provided in the fixture.

**Comment by thedrow** [comment]
FYI
```python
def test_foo(dbsession, foo_facotry):
    foo_facotry._meta.sqlalchemy_session = dbsession
    org = organization_facotry.create()
```

Also works just fine/

**Comment by olegpidsadnyi** [comment]
@thedrow i don't particularly like the idea that _meta parameter is in the model's attribute values, just with the funny name that you probably won't use as an attribute name for your model.

Why did you need another session per factory in the first place? Could you describe your use case?
Are you using 2 sessions in your test code at the same time and why?

**Comment by thedrow** [comment]
No I want to share the same session I created from the pytest fixture with the factories. In order to do that I need to override the session factoryboy uses.

**Comment by olegpidsadnyi** [comment]
@thedrow in this case you didn't need any pull request to factoryboy itself. I think it made it worse because of the parameter pollution.
The _meta.sqlalchemy_session is a lazy callable in the first place.
You could just create an autouse fixture to setup your session as a fixture, that will modify some global variable that is returned by that lazy callable (or defaults to your default settings).

I'd revert that PR in this case. Pytest shouldn't affect how factoryboy works IMO

**Comment by thedrow** [comment]
I don't need or want a global session object.

**Comment by olegpidsadnyi** [comment]
sessionmaker is a factory. so it can return an object that is "global" within your pytest request session.
Which is the case in case of the dbsession fixture.

**Comment by olegpidsadnyi** [comment]
@thedrow are you defining the base factory class?


```python

from sqlalchemy.orm.scoping import scoped_session
from factory.alchemy import SQLAlchemyModelFactory

class ModelFactory(SQLAlchemyModelFactory):
    """Base model factory."""

    class Meta:
        abstract = True
        sqlalchemy_session = scoped_session(<Some callable here that returns a result of your fixture>)

```

**Comment by thedrow** [comment]
No I haven't done that.

**Comment by revmischa** [comment]
Here is what ended up working for me with pytest-flask-sqlalchemy:
```python
Session = scoped_session(
    lambda: current_app.extensions["sqlalchemy"].db.session,
    scopefunc=lambda: current_app.extensions["sqlalchemy"].db.session,
)

class BaseFactory(factory.alchemy.SQLAlchemyModelFactory):
    class Meta:
        abstract = True
        sqlalchemy_session = Session
```
I suggest adding a new subclass of SQLAlchemyModelFactory that contains this setting and plugs automatically into pytest-flask-sqlalchemy without any extra configuration.

**Comment by asduj** [comment]
```Session = scoped_session(
    lambda: current_app.extensions["sqlalchemy"].db.session(),
    scopefunc=lambda: current_app.extensions["sqlalchemy"].db.session().has_key,
)```

**Comment by barraponto** [comment]
To improve on @revmischa solution, I just used db.create_scoped_session provided by Flask-SQLAlchemy:

```python
class FlaskSQLAlchemyModelFactory(factory.alchemy.SQLAlchemyModelFactory):
    """Connects factory meta session to a pytest-flask-sqlalchemy scoped session."""

    @classmethod
    def _setup_next_sequence(cls):
        """Set sequence to start at 1 to keep postgres happy."""
        return 1

    class Meta:  # noqa: D106
        abstract = True
        sqlalchemy_session = db.create_scoped_session()
        sqlalchemy_session_persistence = "commit"

```

**Comment by yobuntu** [comment]
after 1 day working on it, i still don't have a solution, but at least, in can point a problem clearly (i hope):
my problem lies in the `instance become fixtures automatically` (https://pytest-factoryboy.readthedocs.io/en/latest/#integration)

so with the same settings as @revmischa and the fallowing test suite:

```
class TestTest:
    def test_factory(self, user_factory, db_session):
        user = user_factory()

        assert len(User.query.all()) == 1
        assert User.query.all()[0] == user

    def test_factory_rollback(self, db_session):
        assert len(User.query.all()) == 0

    def test_factory_call(self, user, db_session):
        assert len(User.query.all()) == 1

    def test_factory_call_reset(self, db_session):
        assert len(User.query.all()) == 0
```
the first 3 tests pass, and the last one fail impliying there is no rollback on `magic instance` created by the register hook.

here is the factory :
```

Session = scoped_session(
    get_my_session,
    scopefunc=get_my_session,
)


class BaseFactory(SQLAlchemyModelFactory):
    class Meta:
        abstract = True
        sqlalchemy_session = Session
        sqlalchemy_session_persistence = "commit"


@register
class UserFactory(BaseFactory):
    class Meta:
        model = User

    username = factory.Sequence(lambda n: f"user_{n}")
    password = factory.Sequence(
        lambda n: generate_password_hash(f"password_{n}").decode("utf-8")
    )
    email = factory.Sequence(lambda n: f"user_{n}@example.com")
```
